core/rio/turtle/src/main/java/org/eclipse/rdf4j/rio/turtle/TurtleWriter.java

```diff
@@ -129,6 +151,71 @@ public void endRDF()
 		}
 
 		try {
+			if (prettyPrintModel != null) {
+				// Note: We can't guarantee ordering at this point because Resource doesn't implement Comparable<Resource>
+				for (Resource nextContext : prettyPrintModel.contexts()) {
+					for (Resource nextSubject : prettyPrintModel.subjects()) {
+						boolean canShortenSubjectBNode = true;
+						// We can almost always shorten subject IRIs, 
+						// with some known corner cases that are already embedded in the algorithm
+						// So just need to do checking for BNode subjects
+						if (nextSubject instanceof BNode) {
+							if (prettyPrintModel.contains(null, null, nextSubject)) {
+								// Cannot shorten this blank node as it is used as the object of a statement somewhere 
+								// so must be written in a non-anonymous form
+								canShortenSubjectBNode = false;
+							}
```

Comment:
Thinking around future extensions, as you say: it seems to me someway similar to a naive framing algorithm: one could populate a "lazy" collection (stream?) of subject references, then try to resolve them for formatting when needed.
At the same time it could be possible to apply CURIEs or other short representation on the subject uris, as they can improve readability for turtle.